File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        int [] res = new int[1];
        res[0] = Integer.MIN_VALUE;
        rec(root,res);
        return res[0];
    }
    public int rec(TreeNode root,int [] res){
        if(root == null){
            return 0;
        }
        int left = rec(root.left,res);
        int right = rec(root.right,res);
        if(left < 0){
            left = 0;
        }
        if(right < 0){
            right  = 0;
        }
        res[0] = Math.max(res[0], root.val + left + right);
        return root.val + Math.max(left,right);
    }

}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxValue = float('-inf')
        def fun(node):
            nonlocal maxValue
            if node == None:
                return float('-inf')
            left = fun(node.left)
            right = fun(node.right)
            maxValue = max(maxValue, max(max(max(node.val, node.val + left), node.val + right), node.val + left + right))
            return max(node.val, node.val + max(left,right))
        fun(root)
        return maxValue
